clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3710,10 +4348,22 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
                 SYCLIntegrationHeader::kernel_param_kind_t Kind) {
     addParam(ArgTy, Kind, offsetOf(FD, ArgTy));
   }
-  void addParam(QualType ArgTy, SYCLIntegrationHeader::kernel_param_kind_t Kind,
+
+  // For free functions we increment the current offset as each parameter is
+  // added.
+  void addParam(const ParmVarDecl *PD, QualType ParamTy,
+                SYCLIntegrationHeader::kernel_param_kind_t Kind) {
+    addParam(ParamTy, Kind, offsetOf(PD, ParamTy));
+    CurOffset +=
+        SemaSYCLRef.getASTContext().getTypeSizeInChars(ParamTy).getQuantity();
```

Comment:
I'm not sure what this offset is used for. Does this need to account for alignment requirements?

Reply:
The offsets are used for launching normal SYCL kernels. They are not used for free functions but I have kept the values as they might have been for a SYCL kernel.

Reply:
I would have assumed the offsets are required for free functions as well? How are values for decomposed types loaded by runtime without offset information?

Reply:
For pointer and scalar parameters the offset is not needed for free functions. Each parameter is sent to the device individually using Handler.set_arg(<number>,<value>). It is possible that even with decomposed parameters the offsets won't be needed. We'll find out when additional parameter types are implemented. For now, I have kept the offset as it would have been for normal SYCL kernels.